File: backend/app/schemas/rule.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime

class ConditionNode(BaseModel):
    type: str
    op: Optional[str] = None
    field: Optional[str] = None
    value: Optional[Any] = None
    children: Optional[List['ConditionNode']] = None

class RuleBase(BaseModel):
    name: str
    description: Optional[str] = None
    group: Optional[str] = None
    priority: int = 0
    enabled: bool = True
    condition_dsl: Dict[str, Any]
    action: str
    rule_metadata: Optional[Dict[str, Any]] = None

class RuleCreate(RuleBase):
    pass

class RuleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    group: Optional[str] = None
    priority: Optional[int] = None
    enabled: Optional[bool] = None
    condition_dsl: Optional[Dict[str, Any]] = None
    action: Optional[str] = None
    rule_metadata: Optional[Dict[str, Any]] = None

class RuleResponse(RuleBase):
    id: int
    current_version: int
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    
    class Config:
        from_attributes = True

class RuleVersionResponse(BaseModel):
    id: int
    rule_id: int
    version: int
    name: str
    description: Optional[str] = None
    group: Optional[str] = None
    priority: int
    enabled: bool
    condition_dsl: Dict[str, Any]
    action: str
    rule_metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    created_by: Optional[int] = None
    
    class Config:
        from_attributes = True

class SimulateRequest(BaseModel):
    event: Dict[str, Any]
    rule_ids: Optional[List[int]] = None

class SimulateResponse(BaseModel):
    matched_rules: List[Dict[str, Any]]
    execution_order: List[int]
    explanations: Dict[int, str]
    dry_run: bool = True

class DependencyGraph(BaseModel):
    nodes: List[Dict[str, Any]]
    edges: List[Dict[str, Any]]

class ConflictReport(BaseModel):
    conflicts: List[Dict[str, Any]]

File: frontend/src/types/rule.ts
export interface ConditionLeaf {
    type: 'condition';
    field: string;
    op: string;
    value: unknown;
}

export interface ConditionGroup {
    type: 'group';
    op: 'AND' | 'OR';
    children: Array<ConditionGroup | ConditionLeaf>;
}

export type PrimitiveValue = string | number | boolean | null;

export interface IntentPredicate {
    field: string;
    op: string;
    value: PrimitiveValue;
}

export interface IntentPattern {
    intent: string;
    where: IntentPredicate[];
}

export type TimeWindowUnit = 'minutes' | 'hours' | 'days';

export interface TimeWindow {
    value: number;
    unit: TimeWindowUnit;
}

export interface AccumulateDSL {
    type: 'accumulate';
    source_event: string;
    metric_field: string;
    metric_op: '>' | '>=' | '<' | '<=' | '==' | '!=';
    threshold: number;
    window: TimeWindow;
    group_by: string[];
}

export interface SequenceDSL {
    type: 'sequence';
    steps: [IntentPattern, IntentPattern];
    within: TimeWindow;
}

export interface CrossFactJoinDSL {
    type: 'cross_fact_join';
    left: IntentPattern;
    right: IntentPattern;
    join_on: Array<{
        left_field: string;
        right_field: string;
    }>;
    match: 'all' | 'any';
}

export type AdvancedConditionDSL = AccumulateDSL | SequenceDSL | CrossFactJoinDSL;

export type ConditionDSL = ConditionGroup | ConditionLeaf | AdvancedConditionDSL;

export interface Rule {
    id: number;
    name: string;
    description?: string | null;
    group?: string | null;
    priority: number;
    enabled: boolean;
    condition_dsl: ConditionDSL;
    action: string;
    created_at: string;
    updated_at: string;
    current_version: number;
}

export interface RuleCreate {
    name: string;
    description?: string | null;
    group?: string | null;
    priority?: number;
    enabled?: boolean;
    condition_dsl: ConditionDSL;
    action: string;
}

export type RuleUpdate = Partial<RuleCreate>;

export interface RuleVersion {
    version: number;
    created_at: string;
    rule_id: number;
    snapshot?: Record<string, unknown>;
}

export interface RuleVersionDiff {
    rule_id: number;
    version1: number;
    version2: number;
    differences: Record<string, { version1: unknown; version2: unknown }>;
}

export interface MatchedRule {
    id: number;
    name: string;
    action: string;
    priority: number;
    group?: string | null;
}

export interface SimulateResponse {
    matched_rules: MatchedRule[];
    explanations?: Record<string, string>;
    stats?: {
        total_rules: number;
        candidates_evaluated: number;
        evaluation_time_ms: number;
        optimization: string;
    };
}

export interface AvailableAction {
    name: string;
    description: string;
    category?: string;
}

export interface ValidationConflict {
    type: string;
    description: string;
    existing_rule_id?: number;
    existing_rule_name?: string;
}

export interface ValidationResult {
    conflicts: ValidationConflict[];
    similar_rules: Array<{
        rule_id: number;
        rule_name: string;
        similarity_score: number;
        reasons: string[];
        group?: string;
    }>;
}
